src/Skoruba.IdentityServer4.STS.Identity/DependencyInjection/BuilderExtensions.MultiTenant.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Microsoft.AspNetCore.Identity;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using Microsoft.AspNetCore.Hosting;
+using Skoruba.IdentityServer4.Admin.EntityFramework.Shared.DbContexts;
+using Skoruba.IdentityServer4.Admin.EntityFramework.Shared.Entities.Identity;
+using IdentityServer4.EntityFramework.Options;
+using IdentityServer4.Configuration;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Skoruba.IdentityServer4.STS.Identity.DependencyInjection;
+using Skoruba.IdentityServer4.Admin.EntityFramework.Shared.Stores;
+using Skoruba.IdentityServer4.Admin.EntityFramework.Shared.Managers;
+using IdentityServer4.EntityFramework.Interfaces;
+using Skoruba.IdentityServer4.STS.Identity.Services;
+using System.Linq;
+using Skoruba.IdentityServer4.Admin.EntityFramework.Shared.Validators;
+
+namespace Skoruba.IdentityServer4.STS.Identity.DependencyInjection
+{
+    public static partial class BuilderExtensions
+    {
+        public static IBuilder AddMultiTenantConfiguration
+            (this IServiceCollection services, IConfiguration configuration, IHostingEnvironment hostingEnvironment, ILogger logger,
+            Action<DbContextOptionsBuilder> identityDbContextOptions,
+            Action<IdentityOptions> identityOptions,
+            Action<IdentityServerOptions> identityServerOptions,
+            Action<ConfigurationStoreOptions> configurationStoreOptions,
+            Action<OperationalStoreOptions> operationalStoreOptions)
+        {
+            services.AddMemoryCache();
+
+            var builder = services
+                .AddCustomConfiguration
+                    <MultiTenantUserIdentityDbContext, MultiTenantUserIdentity, UserIdentityRole>
+                    (hostingEnvironment, identityDbContextOptions)
+                .AddMultiTenantUserIdentity
+                    <MultiTenantUserIdentityDbContext, MultiTenantUserIdentity, UserIdentityRole>
+                    (identityOptions)
+                .AddMultiTenantIdentityServer
+                    <MultiTenantUserIdentity, IdentityServerConfigurationDbContext, IdentityServerPersistedGrantDbContext>
+                    (configuration, hostingEnvironment, logger, identityServerOptions, configurationStoreOptions, operationalStoreOptions)
+                .AddMvcWithLocalization
+                    <MultiTenantUserIdentity, string>();
+
+            builder.Services.Remove(builder.Services.FirstOrDefault(d => d.ServiceType == typeof(IUserValidator<MultiTenantUserIdentity>)));
+            builder.AddUserValidator<MultiTenantUserIdentity, MultiTenantUserValidator>();
+            builder.AddUserValidator<MultiTenantUserIdentity, RequireTenant>();
```

Comment:
How one could register, should it choose tenant code beforehand like in login form or tenant should be automatically provided per new user?

Reply:
Tenant should be decided beforehand, depending on for which client the user is trying to register.
Usually the subdomain of the client url should provide the info for selecting Tenant. It can also be decided using Request headers.
Example:
Client: xyz.clientA.com and in IS4 client redirect URI: xz.clientA.com/signin-oidc
then process flow would be 
xyz.clientA.com => IS4 - Register UI on register get the tenant code based on subdomain "xyz".